spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskService.java

```diff
@@ -171,75 +179,110 @@ public DefaultTaskService(DataSourceProperties dataSourceProperties,
-					TaskServiceUtils.createComposedTaskDefinition(
-							taskNode.toExecutableDSL(), taskConfigurationProperties));
-			taskDeploymentProperties = TaskServiceUtils.establishComposedTaskProperties(taskDeploymentProperties,
-					taskNode);
-		}
-
-		AppRegistration appRegistration = this.registry.find(taskDefinition.getRegisteredAppName(),
-				ApplicationType.task);
-		Assert.notNull(appRegistration, "Unknown task app: " + taskDefinition.getRegisteredAppName());
-		Resource appResource = this.registry.getAppResource(appRegistration);
-		Resource metadataResource = this.registry.getAppMetadataResource(appRegistration);
-
-		TaskExecution taskExecution = taskExecutionRepository.createTaskExecution(taskName);
-		taskDefinition = TaskServiceUtils.updateTaskProperties(taskDefinition,
-				dataSourceProperties);
-
-		Map<String, String> appDeploymentProperties = new HashMap<>(commonApplicationProperties.getTask());
-		appDeploymentProperties.putAll(
-				TaskServiceUtils.extractAppProperties(taskDefinition.getRegisteredAppName(), taskDeploymentProperties));
-
-		Map<String, String> deployerDeploymentProperties = DeploymentPropertiesUtils
-				.extractAndQualifyDeployerProperties(taskDeploymentProperties, taskDefinition.getRegisteredAppName());
-		if (StringUtils.hasText(this.dataflowServerUri) && taskNode.isComposed()) {
-			TaskServiceUtils.updateDataFlowUriIfNeeded(this.dataflowServerUri, appDeploymentProperties,
-					commandLineArgs);
-		}
-		AppDefinition revisedDefinition = TaskServiceUtils.mergeAndExpandAppProperties(taskDefinition, metadataResource,
-				appDeploymentProperties, this.whitelistProperties);
-		List<String> updatedCmdLineArgs = this.updateCommandLineArgs(commandLineArgs, taskExecution);
-		AppDeploymentRequest request = new AppDeploymentRequest(revisedDefinition, appResource,
-				deployerDeploymentProperties, updatedCmdLineArgs);
-		String id = this.taskLauncher.launch(request);
+		String id = this.taskLauncher.launch(requestCriteria.getAppDeploymentRequest());
 		if (!StringUtils.hasText(id)) {
 			throw new IllegalStateException("Deployment ID is null for the task:" + taskName);
 		}
-		this.taskExecutionRepository.updateExternalExecutionId(taskExecution.getExecutionId(), id);
+		postLaunchProcessing(requestCriteria, id);
+		return requestCriteria.getTaskExecution().getExecutionId();
+	}
+
+	private RequestCriteria createLaunchRequest (final String taskName, final Map<String, String> taskDeploymentProperties,
+			final List<String> commandLineArgs) {
+		TransactionTemplate template = new TransactionTemplate(transactionManager);
+
+		return template.execute(status -> {
+			Assert.hasText(taskName, "The provided taskName must not be null or empty.");
+			Assert.notNull(taskDeploymentProperties, "The provided runtimeProperties must not be null.");
+
+			Map<String, String> launchTaskDeploymentProperties = new HashMap<>(taskDeploymentProperties);
+			if (maxConcurrentExecutionsReached()) {
```

Comment:
This check should move out of this method and be the first line of code in the executeTask method as it isn't related to creating the request.

Reply:
Wouldn't it be safer to keep grouped with the other db calls within the TransactionTemplate?